Networking Metrics (#38422)
dotnet/docs: docs/fundamentals/networking/telemetry/event-counters.md

---
title: Networking Event Counters in .NET
description: Learn how to consume .NET networking event counters.
author: MihaZupan
ms.author: mizupan
ms.date: 10/18/2022
---

# Networking Event Counters in .NET

[EventCounters] are .NET APIs used for lightweight, cross-platform, and near real-time performance metric collection.

Networking components are instrumented to publish basic diagnostics information using EventCounters.
They include information like the following:

- `System.Net.Http` > `requests-started`
- `System.Net.Http` > `requests-failed`
- `System.Net.Http` > `http11-connections-current-total`
- `System.Net.Security` > `all-tls-sessions-open`
- `System.Net.Sockets` > `outgoing-connections-established`
- `System.Net.NameResolution` > `dns-lookups-duration`

> [!TIP]
> For the full list, see [well-known counters].

> 

Reviewer: Add a note that metrics are recommended in .NET 8 or later? And have a link to the metrics page.

---

> title: Networking Event Counters in .NET

Reviewer 2: ```suggestion
title: Networking Event Counters
```

Reviewer 2: A title suffix of ".NET" will be automatically added.

description: Learn how to consume .NET networking event counters.
author: MihaZupan
ms.author: mizupan
ms.date: 10/18/2022
---

> # Networking Event Counters in .NET

Reviewer 2: ```suggestion
# Networking event counters in .NET
```

[EventCounters] are .NET APIs used for lightweight, cross-platform, and near real-time performance metric collection.

> Networking components are instrumented to publish basic diagnostics information using EventCounters.

Reviewer 2: ```suggestion
Networking components are instrumented to publish basic diagnostic information using EventCounters.
```

They include information like the following:

- `System.Net.Http` > `requests-started`
- `System.Net.Http` > `requests-failed`
- `System.Net.Http` > `http11-connections-current-total`
- `System.Net.Security` > `all-tls-sessions-open`
- `System.Net.Sockets` > `outgoing-connections-established`
- `System.Net.NameResolution` > `dns-lookups-duration`

> [!TIP]
> For the full list, see [well-known counters].

> [!TIP]
> On projects targeting .NET 8+, consider using the newer and more feature-rich [networking metrics] instead of EventCounters.

## Providers

Networking information is split across the following providers:

- `System.Net.Http` (`HttpClient` and `SocketsHttpHandler`)
- `System.Net.NameResolution` (`Dns`)
- `System.Net.Security` (`SslStream`)
- `System.Net.Sockets`
- `Microsoft.AspNetCore.Hosting`
- `Microsoft-AspNetCore-Server-Kestrel`

The telemetry has some performance overhead when enabled, so make sure to subscribe only to providers you're actually interested in.

> ## Monitor Event Counters from outside the process

Reviewer 2: ```suggestion
## Monitor event counters from outside the process
```

### dotnet-counters

[`dotnet-counters`][dotnet-counter docs] is a cross-platform performance monitoring tool for ad-hoc health monitoring and first-level performance investigation.

> ```console

Reviewer 2: ````suggestion
```dotnetcli
````

dotnet tool install --global dotnet-counters
```

> ```console

Reviewer 2: ````suggestion
```dotnetcli
````

dotnet-counters monitor --counters System.Net.Http,System.Net.Security --process-id 1234
```

The command continually refreshes the console with the latest numbers.

```txt
[System.Net.Http]
    Current Http 1.1 Connections                       3
    Current Http 2.0 Connections                       1
    Current Http 3.0 Connections                       0
    Current Requests                                   4
    HTTP 1.1 Requests Queue Duration (ms)              0
    HTTP 2.0 Requests Queue Duration (ms)              0
    HTTP 3.0 Requests Queue Duration (ms)              0
    Requests Failed                                    0
    Requests Failed Rate (Count / 1 sec)               0
    Requests Started                                 470
    Requests Started Rate (Count / 1 sec)             18
```

For all the available commands and parameters, see the [dotnet-counter docs].

### Application Insights

Application Insights does not collect event counters by default.
For information on customizing the set of counters you're interested in, see the [AppInsights EventCounters docs].

For example:

> ```c#

Reviewer 2: Can you change all the instances of lowercase `c#` to `C#`? Otherwise it renders as "c#". See https://review.learn.microsoft.com/en-us/dotnet/fundamentals/networking/telemetry/events?branch=pr-en-us-38422#eventlistener.

Author: Done.

services.ConfigureTelemetryModule<EventCounterCollectionModule>((module, options) =>
{
    module.Counters.Add(new EventCounterCollectionRequest("System.Net.Http", "current-requests"));
    module.Counters.Add(new EventCounterCollectionRequest("System.Net.Http", "requests-failed"));
    module.Counters.Add(new EventCounterCollectionRequest("System.Net.Http", "http11-connections-current-total"));
    module.Counters.Add(new EventCounterCollectionRequest("System.Net.Security", "all-tls-sessions-open"));
});
```

> See [this sample](https://github.com/dotnet/docs/tree/main/docs/fundamentals/networking/snippets/misc/RuntimeEventCounters.cs) for subscribing to many runtime and ASP.NET event counters. Simply add an `EventCounterCollectionRequest` for every entry.

Reviewer 2: ```suggestion
For an example of how to subscribe to many runtime and ASP.NET event counters, see the [RuntimeEventCounters sample](https://github.com/dotnet/docs/tree/main/docs/fundamentals/networking/snippets/misc/RuntimeEventCounters.cs). Simply add an `EventCounterCollectionRequest` for every entry.
```

```c#
foreach (var (eventSource, counters) in RuntimeEventCounters.EventCounters)
{
    foreach (string counter in counters)
    {
        module.Counters.Add(new EventCounterCollectionRequest(eventSource, counter));
    }
}
```

> ## Consume Event Counters in-process

Reviewer 2: ```suggestion
## Consume event counters in-process
```

> The [`Yarp.Telemetry.Consumption`] library makes it easy to consume Event Counters from within the process.

Reviewer 2: ```suggestion
The [`Yarp.Telemetry.Consumption`] library makes it easy to consume event counters from within the process.
```
